File: Intatis-Windows/src/Intatis.Core/Providers/OpenAIWireProvider.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json.Nodes;
using Intatis.Core.Protocol;

namespace Intatis.Core.Providers;

// ---------------------------------------------------------------------------
// Tool-calling wire types (shared by Chat and the Code/Cowork agent loops).
// ---------------------------------------------------------------------------

public sealed record ToolSpec
{
    public required string Name { get; init; }
    public required string Description { get; init; }
    public JsonObject Parameters { get; init; } = new()
    {
        ["type"] = "object",
        ["properties"] = new JsonObject(),
    };
}

public sealed record AgentToolCall
{
    public required string Id { get; init; }
    public required string Name { get; init; }
    public required string Arguments { get; init; } // raw JSON string
}

public sealed record AgentMessage
{
    public required string Role { get; init; } // system | user | assistant | tool
    public string? Content { get; init; }
    public List<AgentToolCall>? ToolCalls { get; init; }
    public string? ToolCallId { get; init; }
    public List<ImageAttachment> Images { get; init; } = [];
}

public sealed record AgentRequest
{
    public required string Model { get; init; }
    public required List<AgentMessage> Messages { get; init; }
    public List<ToolSpec> Tools { get; init; } = [];
    public double? Temperature { get; init; }
    public ReasoningEffort? ReasoningEffort { get; init; }
    public bool IncludeUsage { get; init; }
    public bool ParallelToolCalls { get; init; } = false;
}

public abstract record AgentChunk
{
    public sealed record TextDelta(string Text) : AgentChunk;
    public sealed record ToolCallsIssued(List<AgentToolCall> Calls) : AgentChunk;
    public sealed record UsageReport(Usage Usage) : AgentChunk;
    public sealed record Done(string FinishReason) : AgentChunk;
}

public interface IToolCallingProvider
{
    IAsyncEnumerable<AgentChunk> StreamAgentAsync(AgentRequest request, CancellationToken ct = default);
}

// ---------------------------------------------------------------------------
// OpenAI-compatible HTTP adapter (chat/completions, SSE streaming).
// ---------------------------------------------------------------------------

public sealed class OpenAIWireProvider : IChatProvider, IToolCallingProvider
{
    private readonly HttpClient _http;
    private readonly string _baseUrl;
    private readonly string? _chatEndpointOverride;
    private readonly string _apiKey;

    public string BaseUrl => _baseUrl;

    public OpenAIWireProvider(HttpClient http, string baseUrl, string apiKey, string? chatEndpointOverride = null)
    {
        _http = http;
        _baseUrl = baseUrl.TrimEnd('/');
        _chatEndpointOverride = chatEndpointOverride;
        _apiKey = apiKey;
    }

    public static string ChatCompletionsUrl(string baseUrl, string? overrideEndpoint)
        => string.IsNullOrWhiteSpace(overrideEndpoint)
            ? baseUrl.TrimEnd('/') + "/chat/completions"
            : overrideEndpoint;

    public IAsyncEnumerable<ChatChunk> StreamChatAsync(ChatRequest request, CancellationToken ct = default)
        => StreamCoreAsync(ToWire(request), chunk => chunk switch
        {
            AgentChunk.TextDelta delta => (ChatChunk?)new ChatChunk.Delta(delta.Text),
            AgentChunk.UsageReport usage => new ChatChunk.UsageReport(usage.Usage),
            AgentChunk.Done => new ChatChunk.Done(),
            _ => null,
        }, ct);

    public IAsyncEnumerable<AgentChunk> StreamAgentAsync(AgentRequest request, CancellationToken ct = default)
        => StreamCoreAsync(request, chunk => chunk, ct);

    private async IAsyncEnumerable<TOut> StreamCoreAsync<TOut>(
        AgentRequest request,
        Func<AgentChunk, TOut?> map,
        [System.Runtime.CompilerServices.EnumeratorCancellation] CancellationToken ct = default)
    {
        var sawCompletionMarker = false;
        await foreach (var chunk in OpenStream(request, ct).WithCancellation(ct))
        {
            if (chunk is AgentChunk.Done) sawCompletionMarker = true;
            var mapped = map(chunk);
            if (mapped is not null) yield return mapped;
        }
        if (!sawCompletionMarker)
            throw new ProviderException("incomplete_stream", "stream ended without a completion marker");
    }

    private async IAsyncEnumerable<AgentChunk> OpenStream(
        AgentRequest request,
        [System.Runtime.CompilerServices.EnumeratorCancellation] CancellationToken ct = default)
    {
        var url = ChatCompletionsUrl(_baseUrl, _chatEndpointOverride);
        if (!Uri.TryCreate(url, UriKind.Absolute, out var uri) ||
            (uri.Scheme != "http" && uri.Scheme != "https") ||
            string.IsNullOrEmpty(uri.Host))
            throw new ProviderException("config", $"invalid chat endpoint: {url}");

        var body = BuildRequestBody(request);
        using var httpRequest = new HttpRequestMessage(HttpMethod.Post, uri)
        {
            Content = new StringContent(body, Encoding.UTF8, "application/json"),
        };
        httpRequest.Headers.Accept.ParseAdd("text/event-stream");
        if (!string.IsNullOrEmpty(_apiKey))
            httpRequest.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _apiKey);

        using var response = await _http.SendAsync(httpRequest, HttpCompletionOption.ResponseHeadersRead, ct)
            .ConfigureAwait(false);
        if (!response.IsSuccessStatusCode)
        {
            var errorBody = await response.Content.ReadAsStringAsync(ct).ConfigureAwait(false);
            if (errorBody.Length > 2048) errorBody = errorBody[..2048];
            throw new ProviderException("provider.http",
                $"HTTP {(int)response.StatusCode}: {errorBody}", (int)response.StatusCode);
        }

        await using var byteStream = await response.Content.ReadAsStreamAsync(ct).ConfigureAwait(false);
        using var reader = new StreamReader(byteStream, Encoding.UTF8);
        var parser = new SseParser();

        var buffer = new char[8192];
        var toolAccumulator = new Dictionary<int, (string? Id, string Name, StringBuilder Args)>();
        var finishReason = "stop";
        var finishReasonSeen = false;

        int read;
        while ((read = await reader.ReadAsync(buffer, ct).ConfigureAwait(false)) > 0)
        {
            foreach (var dataEvent in parser.Consume(buffer.AsSpan(0, read)))
            {
                if (dataEvent == "[DONE]")
                {
                    yield return new AgentChunk.Done(finishReason);
                    yield break;
                }

                var node = JsonNode.Parse(dataEvent);
                if (node is null) continue;

                var delta = node["choices"]?[0]?["delta"];
                if (delta?["content"] is { } textNode && (string?)textNode is { Length: > 0 } text)
                    yield return new AgentChunk.TextDelta(text);

                if (delta?["tool_calls"] is JsonArray toolCallArray)
                {
                    foreach (var callNode in toolCallArray)
                    {
                        if (callNode is null) continue;
                        var index = (int?)callNode["index"] ?? 0;
                        var id = (string?)callNode["id"];
                        var fn = callNode["function"];
                        var name = (string?)fn?["name"];
                        var argsDelta = (string?)fn?["arguments"];
                        if (!toolAccumulator.TryGetValue(index, out var acc))
                        {
                            acc = (id, name ?? "", new StringBuilder());
                            toolAccumulator[index] = acc;
                        }
                        if (id is not null) acc.Item1 = id;
                        if (name is { Length: > 0 }) acc.Item2 = name;
                        if (argsDelta is { Length: > 0 }) acc.Item3.Append(argsDelta);
                    }
                }

                var reason = (string?)node["choices"]?[0]?["finish_reason"];
                if (reason is { Length: > 0 } && !finishReasonSeen)
                {
                    finishReason = reason;
                    finishReasonSeen = true;
                    foreach (var accumulated in EmitToolCalls(toolAccumulator))
                        yield return accumulated;
                    toolAccumulator.Clear();
                }

                if (node["usage"] is JsonObject usageObj && ParseUsage(usageObj) is { } usage)
                    yield return new AgentChunk.UsageReport(usage);
            }
        }

        foreach (var trailing in parser.Flush())
        {
            if (trailing == "[DONE]") continue;
            var node = JsonNode.Parse(trailing);
            var delta = node?["choices"]?[0]?["delta"];
            if (delta?["content"] is { } textNode && (string?)textNode is { Length: > 0 } text)
                yield return new AgentChunk.TextDelta(text);
        }

        yield return new AgentChunk.Done(finishReason);
    }

    private static IEnumerable<AgentChunk> EmitToolCalls(
        Dictionary<int, (string? Id, string Name, StringBuilder Args)> accumulator)
    {
        if (accumulator.Count == 0) yield break;
        var calls = accumulator.OrderBy(p => p.Key)
            .Where(p => p.Value.Id is not null && p.Value.Name.Length > 0)
            .Select(p => new AgentToolCall
            {
                Id = p.Value.Id!,
                Name = p.Value.Name,
                Arguments = p.Value.Args.ToString(),
            })
            .ToList();
        if (calls.Count > 0)
            yield return new AgentChunk.ToolCallsIssued(calls);
    }

    internal static Usage? ParseUsage(JsonObject obj)
    {
        var prompt = (int?)obj["prompt_tokens"];
        var completion = (int?)obj["completion_tokens"];
        var total = (int?)obj["total_tokens"];
        if (prompt is null && completion is null && total is null) return null;
        return new Usage
        {
            PromptTokens = prompt,
            CompletionTokens = completion,
            TotalTokens = total,
            CachedPromptTokens = (int?)obj["prompt_tokens_details"]?["cached_tokens"],
        };
    }

    /// <summary>Request body bytes are deterministic: sorted keys, images as content parts.</summary>
    internal static string BuildRequestBody(AgentRequest request)
    {
        var body = new JsonObject
        {
            ["model"] = request.Model,
            ["messages"] = new JsonArray(request.Messages.Select(m => (JsonNode?)MessageJson(m)).ToArray()),
            ["stream"] = true,
            ["parallel_tool_calls"] = request.ParallelToolCalls,
        };
        if (request.Temperature is { } temperature) body["temperature"] = temperature;
        if (request.ReasoningEffort is { } effort) body["reasoning_effort"] = effort.ToWire();
        if (request.IncludeUsage)
            body["stream_options"] = new JsonObject { ["include_usage"] = true };
        if (request.Tools is { Count: > 0 } tools)
            body["tools"] = new JsonArray(tools.Select(t => (JsonNode?)new JsonObject
            {
                ["type"] = "function",
                ["function"] = new JsonObject
                {
                    ["name"] = t.Name,
                    ["description"] = t.Description,
                    ["parameters"] = t.Parameters.DeepClone(),
                },
            }).ToArray());

        return Jsonx.SerializeSorted(body);
    }

    private static JsonObject MessageJson(AgentMessage message)
    {
        if (message.Images.Count > 0 && message.Role == "user")
        {
            var parts = new JsonArray();
            if (!string.IsNullOrEmpty(message.Content))
                parts.Add(new JsonObject { ["type"] = "text", ["text"] = message.Content });
            foreach (var image in message.Images)
                parts.Add(new JsonObject
                {
                    ["type"] = "image_url",
                    ["image_url"] = new JsonObject { ["url"] = image.Url },
                });
            return new JsonObject { ["role"] = message.Role, ["content"] = parts };
        }

        var obj = new JsonObject { ["role"] = message.Role };
        if (message.ToolCalls is { Count: > 0 })
        {
            obj["content"] = string.IsNullOrEmpty(message.Content) ? null : message.Content;
            obj["tool_calls"] = new JsonArray(message.ToolCalls.Select(c => (JsonNode?)new JsonObject
            {
                ["id"] = c.Id,
                ["type"] = "function",
                ["function"] = new JsonObject
                {
                    ["name"] = c.Name,
                    ["arguments"] = c.Arguments,
                },
            }).ToArray());
        }
        else
        {
            obj["content"] = message.Content ?? "";
        }
        if (message.Role == "tool" && message.ToolCallId is not null)
            obj["tool_call_id"] = message.ToolCallId;
        return obj;
    }

    private static AgentRequest ToWire(ChatRequest request) => new()
    {
        Model = request.Model,
        Messages = request.Messages
            .Select(m => new AgentMessage { Role = m.Role, Content = m.Content, Images = m.Images })
            .ToList(),
        Temperature = request.Temperature,
        ReasoningEffort = request.ReasoningEffort,
        IncludeUsage = request.IncludeUsage,
    };
}

File: Intatis-Android/shared/src/main/kotlin/com/intatis/shared/providers/OpenAIWireProvider.kt
package com.intatis.shared.providers

import com.intatis.shared.protocol.Jsonx
import com.intatis.shared.protocol.Jsonx.int
import com.intatis.shared.protocol.Jsonx.str
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

/**
 * OpenAI-compatible HTTP adapter (chat/completions, SSE streaming). The mobile app is
 * a strict Chat subset per the Apple contract — no tool calling on this surface.
 */
class OpenAIWireProvider(
    private val client: OkHttpClient,
    private val baseUrl: String,
    private val apiKey: String,
    private val chatEndpointOverride: String? = null,
) : ChatProvider {

    val effectiveBaseUrl: String get() = baseUrl.trimEnd('/')

    companion object {
        fun chatCompletionsUrl(baseUrl: String, override: String?): String =
            if (override.isNullOrBlank()) baseUrl.trimEnd('/') + "/chat/completions"
            else override

        fun defaultClient(): OkHttpClient = OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(300, TimeUnit.SECONDS)
            .build()

        internal fun parseUsage(obj: JsonObject): Usage? {
            val prompt = obj.int("prompt_tokens")
            val completion = obj.int("completion_tokens")
            val total = obj.int("total_tokens")
            if (prompt == null && completion == null && total == null) return null
            val cached = (obj["prompt_tokens_details"] as? JsonObject)?.int("cached_tokens")
            return Usage(prompt, cached, completion, total)
        }
    }

    override fun streamChat(request: ChatRequest): Flow<ChatChunk> = flow {
        val url = chatCompletionsUrl(baseUrl, chatEndpointOverride)
        val parsed = java.net.URI(url)
        require(parsed.scheme == "http" || parsed.scheme == "https") { "invalid chat endpoint: $url" }
        require(parsed.host?.isNotEmpty() == true) { "invalid chat endpoint: $url" }

        val body = buildRequestBody(request)
        val httpRequest = Request.Builder()
            .url(url)
            .post(body.toRequestBody("application/json".toMediaType()))
            .header("Accept", "text/event-stream")
            .apply {
                if (apiKey.isNotEmpty()) header("Authorization", "Bearer $apiKey")
            }
            .build()

        client.newCall(httpRequest).execute().use { response ->
            if (!response.isSuccessful) {
                var errorBody = response.body?.string() ?: ""
                if (errorBody.length > 2048) errorBody = errorBody.take(2048)
                throw ProviderException(
                    "provider.http",
                    "HTTP ${response.code}: $errorBody",
                    response.code,
                )
            }

            val source = response.body?.source()
                ?: throw ProviderException("provider.http", "empty response body", response.code)
            val parser = SseParser()
            var finishReasonSeen = false
            var sawDone = false

            // OkHttp reads are blocking; the flow is pinned to IO below.
            val reader = java.io.BufferedReader(source.inputStream().reader(Charsets.UTF_8))
            while (true) {
                val line = reader.readLine() ?: break
                val chunk = if (line.isEmpty()) "" else line + "\n"
                for (event in parser.consume(chunk)) {
                    if (event == "[DONE]") {
                        sawDone = true
                        emit(ChatChunk.Done)
                        return@use
                    }
                    val node = try {
                        Jsonx.lenient.parseToJsonElement(event)
                    } catch (_: Exception) {
                        null
                    } ?: continue

                    val delta = (node as? JsonObject)
                        ?.get("choices")?.let { it as? JsonArray }
                        ?.firstOrNull()?.let { it as? JsonObject }
                        ?.get("delta")?.let { it as? JsonObject }
                    val text = delta?.str("content")
                    if (!text.isNullOrEmpty()) emit(ChatChunk.Delta(text))

                    val finishReason = (node as? JsonObject)
                        ?.get("choices")?.let { it as? JsonArray }
                        ?.firstOrNull()?.let { it as? JsonObject }
                        ?.str("finish_reason")
                    if (!finishReason.isNullOrEmpty()) finishReasonSeen = true

                    (node as? JsonObject)?.get("usage")?.let { it as? JsonObject }?.let { usageObj ->
                        parseUsage(usageObj)?.let { emit(ChatChunk.UsageReport(it)) }
                    }
                }
            }

            if (finishReasonSeen || sawDone) {
                emit(ChatChunk.Done)
            } else {
                throw ProviderException("incomplete_stream", "stream ended without a completion marker")
            }
        }
    }.flowOn(Dispatchers.IO)

    internal fun buildRequestBody(request: ChatRequest): String {
        val body = buildJsonObject {
            put("model", request.model)
            put("messages", buildJsonArray {
                request.messages.forEach { message ->
                    add(messageJson(message))
                }
            })
            put("stream", true)
            request.temperature?.let { put("temperature", it) }
            request.reasoningEffort?.let { put("reasoning_effort", it.wire) }
            if (request.includeUsage) {
                put("stream_options", buildJsonObject { put("include_usage", true) })
            }
        }
        return Jsonx.serializeSorted(body)
    }

    private fun messageJson(message: ChatMessage): JsonObject {
        if (message.images.isNotEmpty() && message.role == "user") {
            return buildJsonObject {
                put("role", message.role)
                put("content", buildJsonArray {
                    if (message.content.isNotEmpty()) {
                        add(buildJsonObject {
                            put("type", "text")
                            put("text", message.content)
                        })
                    }
                    message.images.forEach { image ->
                        add(buildJsonObject {
                            put("type", "image_url")
                            put("image_url", buildJsonObject { put("url", image.url) })
                        })
                    }
                })
            }
        }
        return buildJsonObject {
            put("role", message.role)
            put("content", message.content)
        }
    }
}
